### PR TITLE
Rewrite internal `jobs` API

### PR DESCRIPTION
**Pull request by infomiho**
### Motivation

We wanted to directly import user code when registering jobs so the error reporting would stay correct. But we also needed to expose an API for users to use jobs in their server code from SDK. This prompted a bit of a rewrite of our internal jobs API.

Let's see how it looks now.

### 1️⃣ SDK

In the SDK, we generate a "job definition" that contains the name, default params, schedule etc. (all of the details) for a job. This is enough for the user to import their jobs in e.g. their actions and submit the job to PgBoss for execution.

```ts
import { prisma } from 'wasp/server'
import type { JSONValue, JSONObject } from 'wasp/server/_types/serialization'
import { type JobFn, createJobDefinition } from 'wasp/server/jobs/core/pgBoss'

const entities = {
  Task: prisma.task,
}

// PUBLIC API
export type PrintTimeAndNumberOfTasks<Input extends JSONObject, Output extends JSONValue | void> = JobFn<Input, Output, typeof entities>

// PUBLIC API
export const printTimeAndNumberOfTasks = createJobDefinition({
  jobName: 'printTimeAndNumberOfTasks',
  defaultJobOptions: {},
  jobSchedule: {"cron":"* * * * *","options":{},"args":null},
  entities,
})
```

### 2️⃣ Server

In the server, we import the "job definition" from the SDK and the user defined function to register a job handler with PgBoss. This way we keep the info about the job in one place but we are able to import user's code directly ✅ 

```ts
import { registerJob } from 'wasp/server/jobs/core/pgBoss'
import { printTimeAndNumberOfTasks } from '../../../../../src/jobs/print'
import { printTimeAndNumberOfTasks as _waspJobDefinition } from 'wasp/server/jobs'

registerJob({
  job: _waspJobDefinition,
  jobFn: printTimeAndNumberOfTasks,
})
```

More details: https://www.notion.so/wasp-lang/Jobs-API-w-SDK-64dff5dc719544369786fe8661a3793b?pvs=4

### CODE REVIEW

File: waspc/src/Wasp/Generator/SdkGenerator/Server/JobGenerator.hs [hunk 1]
```diff
@@ -42,29 +52,81 @@ genIndexTs jobs = return $ C.mkTmplFdWithData tmplFile tmplData
...
-          "jobExecutorTypesImportPath" .= SP.fromRelFileP jobExecutorTypesImportPath,
-          "entities" .= maybe [] (map (makeJsonWithEntityData . AS.refName)) (J.entities job)
+        [ "jobName" .= jobName,
+          "typeName" .= toUpperFirst jobName,
+          "jobExecutorImportPath" .= SP.fromRelFileP jobExecutorImportPath,
+          "entities" .= maybe [] (map (makeJsonWithEntityData . AS.refName)) (J.entities job),
+          -- NOTE: You cannot directly input an Aeson.object for Mustache to substitute.
+          -- This is why we must get a text representation of the object, either by
+          -- `Aeson.Text.encodeToLazyText` on an Aeson.Object, or `show` on an AS.JSON.
+          "jobSchedule" .= Aeson.Text.encodeToLazyText (fromMaybe Aeson.Null maybeJobSchedule),
```
Comment (sodic) [inline waspc/src/Wasp/Generator/SdkGenerator/Server/JobGenerator.hs:74]: Hm, I'm not sure about this. I remember doing it for generating query and action files. We usually send `Aeson.Value`, but those can be objects too, right? I'm not sure what the difference is.

**Review by sodic — APPROVED** [review]
Nice work figuring this out.

I understand the trade-off, but I think it's both worth it and our only option.

File: waspc/data/Generator/templates/sdk/server/jobs/core/pgBoss/pgBossJob.ts [hunk 2]
```diff
@@ -3,40 +3,61 @@ import { pgBossStarted } from './pgBoss.js'
...
+  )
+}
+
+// PRIVATE API
+/**
+ * Uses the info about a job in PgBossJob to register a user defined job handler with pg-boss.
+ * We expect this to be called once per job name. If called multiple times with the same name and different
+ * functions, we will override the previous calls.
+ */
+export function registerJob<
```
[the remark this answers is not in the document]
Comment (infomiho) [reply waspc/data/Generator/templates/sdk/server/jobs/core/pgBoss/pgBossJob.ts:54]: We need stuff for submitting jobs in the SDK, to be more specific the `pgBossStarted` symbol which gives us the `boss` instance of PgBoss. Then, we can call `boss.send`.

Okay, so to have `pgBossStarted` in the SDK, we need to have `startPgBoss` in the SDK. Which means that we need to basically have everything related to PgBoss in the SDK.

`registerJob` also needs the `boss` instance to register the jobs, so yep, it also needs to be in the SDK.

File: waspc/data/Generator/templates/server/src/jobs/_job.ts [hunk 3]
```diff
@@ -1,12 +1,9 @@
...
+{=& jobPerformFn.importStatement =}
+{=& jobDefinition.importStatement =}
 
-export const {= jobName =} = createJob({
-  jobName: "{= jobName =}",
-  jobFn: {= jobPerformFnName =},
-  defaultJobOptions: {=& jobPerformOptions =},
-  jobSchedule: {=& jobSchedule =},
-  {= jobEntitiesIdentifier =},
+registerJob({
```
[the remark this answers is not in the document]
Comment (infomiho) [reply waspc/data/Generator/templates/server/src/jobs/_job.ts:6]: > I'm guessing it's to keep changes to a minimum.

Yep, I wanted to rewrite things as little as possible.

File: waspc/src/Wasp/Generator/SdkGenerator/Server/JobGenerator.hs (hunk 1 above)
Comment (infomiho) [reply waspc/src/Wasp/Generator/SdkGenerator/Server/JobGenerator.hs:74]: > Hm, I'm not sure about this. I remember doing it for generating query and action files. We usually send Aeson.Value, but those can be objects too, right? I'm not sure what the difference is.

This is old code and I didn't want to change it alongside all the other changes in this PR. We can rewrite this later, but it also works okay now 🤷

**Pull request merged**